src/main/java/intellistart/interviewplanning/model/slot/SlotRepository.kt

```diff
@@ -0,0 +1,82 @@
+package intellistart.interviewplanning.model.slot
+
+import intellistart.interviewplanning.model.user.Candidate
+import intellistart.interviewplanning.model.user.Interviewer
+import intellistart.interviewplanning.model.user.User
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.core.FindAndModifyOptions
+import org.springframework.data.mongodb.core.MongoTemplate
+import org.springframework.data.mongodb.core.aggregation.Aggregation
+import org.springframework.data.mongodb.core.query.Criteria
+import org.springframework.data.mongodb.core.query.Query
+import org.springframework.data.mongodb.core.query.Update
+import org.springframework.stereotype.Repository
+import java.time.LocalDate
+
+@Repository
+class SlotRepository(private val mongoTemplate: MongoTemplate) {
+
+    fun findByEmail(email: String): List<Slot> {
+        val aggregation = Aggregation.newAggregation(
+            Aggregation.match(Criteria.where("email").`is`(email)),
+            Aggregation.unwind("slots"),
+            Aggregation.replaceRoot("slots")
+        )
+
+        val results = mongoTemplate.aggregate(aggregation, User.COLLECTION_NAME, Slot::class.java)
+        return results.mappedResults
+    }
+
+    fun findById(id: ObjectId): Slot? {
+        val query = Query(Criteria.where("slots._id").`is`(id))
+        return when (val user = mongoTemplate.findOne(query, User::class.java)) {
+            is Candidate -> user.slots.find { it.id == id }
+            is Interviewer -> user.slots.find { it.id == id }
+            else -> null
```

Comment:
try avoid duplicate user.slots.find { it.id == id }

Reply:
I have one Booking stored in two places - one in the candidate slot and one in the interviewer slot. So as a result of the id search I will have two identical bookings - that is an array, and I take one of these two identical records using .find { it.id == id }

Reply:
So, you fetch from the DB complete user document but you need only user's slot object.

It looks like you need **projection in the query**. Fetch from the DB only data that you require.

Reply:
Ok, consider my comment as optional. Let's move on.

Note: the comment from @mykolaKhlopyk is reasonable.